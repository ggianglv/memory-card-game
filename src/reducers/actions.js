export const SET_MODE = "SET_MODE";
export const UPDATE_TIME = "UPDATE_TIME";
export const UPDATE_MOVE = "UPDATE_MOVE";
export const UPDATE_DATA = "SET_INIT_DATA";
export const UPDATE_START = "UPDATE_START";
export const UPDATE_TOPIC = "UPDATE_TOPIC";
export const UPDATE_REMOVED = "UPDATE_REMOVED";
