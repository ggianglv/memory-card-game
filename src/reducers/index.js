import { LEVEL, TOPICS } from "../constants";
import {
  UPDATE_DATA,
  SET_MODE,
  UPDATE_MOVE,
  UPDATE_REMOVED,
  UPDATE_START,
  UPDATE_TIME,
  UPDATE_TOPIC,
} from "./actions";
import { generateGameData } from "../utils/data";

const initData = generateGameData({ topic: TOPICS.RICK, mode: LEVEL.EASY });

export const initialState = {
  move: 0,
  removed: {},
  time: 0,
  start: false,
  topic: TOPICS.RICK,
  mode: LEVEL.EASY,
  isLoading: false,
  complete: false,
  history: [],
  ...initData,
};

export const appReducer = (state, action) => {
  switch (action.type) {
    case SET_MODE:
      const data = generateGameData({
        topic: state.topic,
        mode: action.payload,
      });
      return {
        ...state,
        ...data,
        mode: action.payload,
      };
    case UPDATE_DATA:
      return {
        ...state,
        ...action.payload,
      };
    case UPDATE_START:
      return {
        ...state,
        start: action.payload,
      };

    case UPDATE_MOVE:
      return {
        ...state,
        move: action.payload,
      };
    case UPDATE_TIME:
      return {
        ...state,
        time: state.time + 1,
      };
    case UPDATE_REMOVED:
      return {
        ...state,
        removed: action.payload,
      };
    case UPDATE_TOPIC: {
      const data = generateGameData({
        topic: action.payload,
        mode: state.mode,
      });
      return {
        ...state,
        ...data,
        topic: action.payload,
        start: false,
        time: 0,
        move: 0,
      };
    }
    default:
      return state;
  }
};
