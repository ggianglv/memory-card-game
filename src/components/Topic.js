import React, { useContext } from "react";
import { AppContext } from "../context";
import { UPDATE_TOPIC } from "../reducers/actions";

const Topic = ({ closeModal }) => {
  const { state, dispatch } = useContext(AppContext);

  const changeTopic = (topic) => {
    dispatch({
      type: UPDATE_TOPIC,
      payload: topic,
    });
    closeModal();
  };

  const topics = [
    {
      key: "pepe",
      name: "Pepe",
    },
    {
      key: "rick-and-morty",
      name: "Rick & morty",
    },
  ];

  return (
    <div className="topic">
      {topics.map((topic) => (
        <div
          onClick={() => changeTopic(topic.key)}
          key={topic.key}
          data-testid={topic.key}
          className={`modal-item ${topic.key === state.topic ? "active" : ""}`}
        >
          {topic.name}
        </div>
      ))}
    </div>
  );
};

export default Topic;
