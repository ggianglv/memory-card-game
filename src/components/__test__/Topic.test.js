import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render } from "@testing-library/react";
import { AppContext } from "../../context";
import { initialState } from "../../reducers";
import Topic from "../Topic";
import renderer from "react-test-renderer";

test("Topic render correctly", () => {
  const { getByTestId } = render(
    <AppContext.Provider value={{ state: initialState, dispatch: () => {} }}>
      <Topic />
    </AppContext.Provider>
  );
  const topic = initialState.topic;

  expect(getByTestId(topic).classList.contains("active")).toBe(true);
});

test("Topic to match snapshot", () => {
  const tree = renderer
    .create(
      <AppContext.Provider value={{ state: initialState, dispatch: () => {} }}>
        <Topic />
      </AppContext.Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
