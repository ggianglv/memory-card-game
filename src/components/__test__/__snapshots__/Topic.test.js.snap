// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Topic to match snapshot 1`] = `
<div
  className="topic"
>
  <div
    className="modal-item "
    data-testid="pepe"
    onClick={[Function]}
  >
    Pepe
  </div>
  <div
    className="modal-item active"
    data-testid="rick-and-morty"
    onClick={[Function]}
  >
    Rick & morty
  </div>
</div>
`;
